### PR TITLE
 Added request parameters through URL

### PR DESCRIPTION
**Pull request by hod-alpert**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/23464

### New argument
- the output format of the exported indicators is - panosurl

- n - The maximum number of entries in the output. If no value is provided, will use the value specified in the List Size parameter configured in the instance configuration.
Example: https://{demisto_instance}/instance/execute/{ExportIndicators_instance_name}?n=50

- s - The starting entry index from which to export the indicators.
Example: https://{demisto_instance}/instance/execute/{ExportIndicators_instance_name}?s=10&n=50

- q - The query used to retrieve indicators from the system.
Example: https://{demisto_instance}/instance/execute/{ExportIndicators_instance_name}?q=“type:ip and sourceBrand:my_source”

- sp -  If set will strip ports off URLs, otherwise will ignore URLs with ports.
https://{demisto_instance}/instance/execute/{ExportIndicators_instance_name}?q=“type:ip and sourceBrand:my_source”&sp

 - di -   If set will ignore URLs which are not compliant with PAN-OS URL format instead of being re-written.
https://{demisto_instance}/instance/execute/{ExportIndicators_instance_name}?q=“type:ip and sourceBrand:my_source”&di

- tr - Whether to collapse IPs.
0 - Do not collapse.
1 - Collapse to ranges.
2 - Collapse to CIDRs.
Example: https://{demisto_instance}/instance/execute/{ExportIndicators_instance_name}?q=“type:ip and sourceBrand:my_source”&tr=1

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [x] 5.5.0

## Does it break backward compatibility?
   - [x] Yes
       - Further details:
   - [ ] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 1]
```diff
@@ -431,26 +476,53 @@ def route_edl_values() -> Response:
...
+        last_update_data=demisto.getIntegrationContext(),
         cache_refresh_rate=params.get('cache_refresh_rate'),
-        panos_compatible=panos_compatible,
-        url_port_stripping=url_port_stripping,
-        collapse_ips=params.get('collapse_ips')
     )
     return Response(values, status=200, mimetype='text/plain')
 
 
+def get_request_args(params):
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:491]: add docstring and type hints

Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:491]: pass `request.args` as arg

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 2]
```diff
@@ -431,26 +476,53 @@ def route_edl_values() -> Response:
...
+    limit = try_parse_integer(request.args.get('n', params.get('edl_size', 10000)), EDL_LIMIT_ERR_MSG)
+    offset = try_parse_integer(request.args.get('s', 0), EDL_OFFSET_ERR_MSG)
+    query = request.args.get('q', params.get('indicators_query'))
+    strip_port = request.args.get('sp', params.get('url_port_stripping', False))
+    drop_invalids = request.args.get('di', params.get('drop_invalids', False))
+    collapse_ips = request.args.get('tr', params.get('collapse_ips', DONT_COLLAPSE))
+
+    # handle flags
+    if drop_invalids is not None and drop_invalids == '':
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:500]:
```suggestion
    if drop_invalids == '':
```

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 3]
```diff
@@ -431,26 +476,53 @@ def route_edl_values() -> Response:
...
+        drop_invalids = True
+
+    if strip_port is not None and strip_port == '':
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:503]:
```suggestion
    if strip_port == '':
```

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 4]
```diff
@@ -431,26 +476,53 @@ def route_edl_values() -> Response:
...
+        strip_port = True
+
+    if collapse_ips is not None and collapse_ips not in [DONT_COLLAPSE, COLLAPSE_TO_CIDR, COLLAPSE_TO_RANGES]:
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:506]:
```suggestion
    if collapse_ips not in [DONT_COLLAPSE, COLLAPSE_TO_CIDR, COLLAPSE_TO_RANGES]:
```

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 5]
```diff
@@ -313,70 +314,114 @@ def ips_to_ranges(ips: list, collapse_ips):
...
+        ipv6_formatted_indicators = ips_to_ranges(ipv6_formatted_indicators, request_args.collapse_ips)
         formatted_indicators.extend(ipv6_formatted_indicators)
-
     return {EDL_VALUES_KEY: list_to_str(formatted_indicators, '\n')}, len(formatted_indicators)
 
 
-def get_edl_ioc_values(on_demand, limit, indicator_query='', last_run=None, cache_refresh_rate=None,
-                       collapse_ips: str = DONT_COLLAPSE, panos_compatible: bool = True,
-                       url_port_stripping: bool = False) -> str:
+def get_edl_ioc_values(on_demand: bool,
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:374]: add type hints and docstring

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 6]
```diff
@@ -313,70 +314,114 @@ def ips_to_ranges(ips: list, collapse_ips):
...
+                       request_args: RequestArguments,
+                       last_update_data={},
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:376]:
```suggestion
                       last_update_data,
```

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 7]
```diff
@@ -313,70 +314,114 @@ def ips_to_ranges(ips: list, collapse_ips):
...
+                       cache_refresh_rate=None) -> str:
     """
     Get the ioc list to return in the edl
     """
+    if last_update_data is None:
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:381]: remove

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 8]
```diff
@@ -313,70 +314,114 @@ def ips_to_ranges(ips: list, collapse_ips):
...
-            values_str = refresh_edl_context(indicator_query, limit=limit,
-                                             panos_compatible=panos_compatible,
-                                             url_port_stripping=url_port_stripping,
-                                             collapse_ips=collapse_ips)
+            values_str = refresh_edl_context(request_args)
     return values_str
 
 
-def get_ioc_values_str_from_context() -> str:
+def get_ioc_values_str_from_context(request_args: RequestArguments, iocs=None) -> str:
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:407]: add type hint and docstring

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 9]
```diff
@@ -313,70 +314,114 @@ def ips_to_ranges(ips: list, collapse_ips):
...
     """
-    cache_dict = demisto.getIntegrationContext()
-    return cache_dict.get(EDL_VALUES_KEY, '')
+    if iocs:
+        if request_args.offset > len(iocs):
+            return ''
+
+        iocs = iocs[request_args.offset: request_args.limit + request_args.offset]
+        returned_dict, _ = create_values_for_returned_dict(iocs, request_args=request_args)
+        current_cache = demisto.getIntegrationContext()
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:417]: dont call `getIntegrationContext`, pass it from the function that call this one

File: Packs/EDL/Integrations/EDL/EDL_description.md [hunk 10]
```diff
@@ -14,3 +14,26 @@ To access the EDL service by instance name, make sure ***Instance execute extern
...
+5. **di** - If set will ignore URLs which are not compliant with PAN-OS URL format instead of being re-written.
+ * Example: https://{cortex-xsoar_instance}/instance/execute/{EDL_instance_name}?v=panosurl&di
+6. **tr** - Whether to collapse IPs. 
+    * 0 - Do not collapse. 
+    * 1 - Collapse to ranges.
+    * 2 - Collapse to CIDRs.
+ * Example: https://{cortex-xsoar_instance}/instance/execute/{ExportIndicators_instance_name}?q="type:ip and sourceBrand:my_source"&tr=1
+
+### When running in On-Demand mode
+Please make sure to to run !eis-update for the first time to initialize the export process.
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL_description.md:39]:
```suggestion
Please make sure to to run !edl-update for the first time to initialize the export process.
```

File: Packs/EDL/Integrations/EDL/EDL_description.md [hunk 11]
```diff
@@ -14,3 +14,26 @@ To access the EDL service by instance name, make sure ***Instance execute extern
...
+ * Example: https://{cortex-xsoar_instance}/instance/execute/{EDL_instance_name}?q="type:ip and sourceBrand:my_source"
+4. **sp** - If set will strip ports off URLs, otherwise will ignore URLs with ports.
+ * Example: https://{cortex-xsoar_instance}/instance/execute/{EDL_instance_name}?v=panosurl&sp 
+5. **di** - If set will ignore URLs which are not compliant with PAN-OS URL format instead of being re-written.
+ * Example: https://{cortex-xsoar_instance}/instance/execute/{EDL_instance_name}?v=panosurl&di
+6. **tr** - Whether to collapse IPs. 
+    * 0 - Do not collapse. 
+    * 1 - Collapse to ranges.
+    * 2 - Collapse to CIDRs.
+ * Example: https://{cortex-xsoar_instance}/instance/execute/{ExportIndicators_instance_name}?q="type:ip and sourceBrand:my_source"&tr=1
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL_description.md:36]:
```suggestion
 * Example: https://{cortex-xsoar_instance}/instance/execute/{EDL_instance_name}?q="type:ip and sourceBrand:my_source"&tr=1
```

File: Packs/EDL/Integrations/EDL/CHANGELOG.md [hunk 12]
```diff
@@ -1,4 +1,5 @@
 ## [Unreleased]
+  - Added Request Parameters Through the URL
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/CHANGELOG.md:2]: add breaking change note

File: Packs/EDL/Integrations/EDL/CHANGELOG.md [hunk 13]
```diff
@@ -1,4 +1,6 @@
 ## [Unreleased]
+  - Removed panos_compatible parameter, all indicators exported by this integrations will be panos compatible.
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/CHANGELOG.md:2]: @kirbles19 note that this is a breaking change

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 14]
```diff
@@ -28,18 +26,59 @@ def write(msg):
...
+_INVALID_TOKEN_REMOVAL = re.compile(r'(?:[^\./+=\?&]+\*[^\./+=\?&]*)|(?:[^\./+=\?&]*\*[^\./+=\?&]+)')
 
 DONT_COLLAPSE = "Don't Collapse"
 COLLAPSE_TO_CIDR = "To CIDRS"
 COLLAPSE_TO_RANGES = "To Ranges"
 
+'''Request Arguments Class'''
+
+
+class RequestArguments:
```
Comment (DeanArbel) [inline Packs/EDL/Integrations/EDL/EDL.py:47]: @roysagi would you mind taking a look at this?

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 15]
```diff
@@ -313,70 +315,127 @@ def ips_to_ranges(ips: list, collapse_ips):
...
-            values_str = refresh_edl_context(indicator_query, limit=limit,
-                                             panos_compatible=panos_compatible,
-                                             url_port_stripping=url_port_stripping,
-                                             collapse_ips=collapse_ips)
+            values_str = refresh_edl_context(request_args)
     return values_str
 
 
-def get_ioc_values_str_from_context() -> str:
+def get_ioc_values_str_from_context(integration_context: dict, request_args: RequestArguments, iocs=None) -> str:
```
Comment (Itay4) [inline Packs/EDL/Integrations/EDL/EDL.py:415]: add type hint for `iocs`

**Comment by kirbles19** [comment]
@hod-alpert Done.

**Pull request merged**